Fix requires_lto targets needing lto set in cargo

Flakebi opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149624)*
<!-- homu-ignore:end -->

Targets that set `requires_lto = true` were not actually using lto when compiling with cargo by default. They needed an extra `lto = true` in `Cargo.toml` to work.

Fix this by letting lto take precedence over the `embed_bitcode` flag when lto is required by a target.

If both these flags would be supplied by the user, an error is generated. However, this did not happen when lto was requested by the target instead of the user.

Fixes rust-lang/rust#148514
Tracking issue: rust-lang/rust#135024

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19914525071/job/57090436076) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/57090436076) [(plain)](https://github.com/rust-lang/rust/commit/a12ddcf1eccc5826c8a9f311fb15527c8868b6e4/checks/57090436076/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6589 files
tidy check
tidy [target-specific-tests (tests)]: /checkout/tests/run-make-cargo/amdgpu-lto/rmake.rs: revision [unspecified] should not specify `needs-llvm-components:` as it doesn't need `--target`
tidy [target-specific-tests (tests)]: FAIL
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
---
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: target-specific-tests (tests)
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/yarn --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1612:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1317:29

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:03:00
  local time: Thu Dec  4 01:34:01 UTC 2025
  network time: Thu, 04 Dec 2025 01:34:01 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

jieyouxu commented:
> The CI/tidy complain seems like a false-positive, the test uses `--target`, just not through `//@ compile-flags:`. Maybe that lint shouldn’t run on `run-make-cargo` tests? (I’m happy to make the change, just want to get confirmation that it’s ok to do that.)

Yes, I probably forgot to extend the tidy exception to `run-make-cargo` when I split the previous run-make test suite.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/write.rs:
```diff
@@ -410,7 +410,7 @@ struct CompiledModules {
 fn need_bitcode_in_object(tcx: TyCtxt<'_>) -> bool {
     let sess = tcx.sess;
-    sess.opts.cg.embed_bitcode
+    (sess.lto() != config::Lto::No || sess.opts.cg.embed_bitcode)
```
Indeed. thin-local LTO shouldn't emit bitcode.

Mark-Simulacrum replied:
Yeah, perf results seem to confirm this being not what we want :)

rust-timer commented:
Queued d1b36b73042e93fe4f006b06efda8d0018ccc5bd with parent 9f54abeebae58244d0d1c692af6f37d2b521e136, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9f54abeebae58244d0d1c692af6f37d2b521e136&end=d1b36b73042e93fe4f006b06efda8d0018ccc5bd).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (d1b36b73042e93fe4f006b06efda8d0018ccc5bd): [comparison URL](https://perf.rust-lang.org/compare.html?start=9f54abeebae58244d0d1c692af6f37d2b521e136&end=d1b36b73042e93fe4f006b06efda8d0018ccc5bd&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=9f54abeebae58244d0d1c692af6f37d2b521e136&end=d1b36b73042e93fe4f006b06efda8d0018ccc5bd&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3% | [0.1%, 3.6%] | 29    |
| Regressions ❌ <br /> (secondary)  | 1.7% | [0.2%, 3.6%] | 26    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.3% | [0.1%, 3.6%] | 29    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9f54abeebae58244d0d1c692af6f37d2b521e136&end=d1b36b73042e93fe4f006b06efda8d0018ccc5bd&stat=max-rss)
<details>
<summary>Results (primary 0.1%, secondary 2.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.9%  | [2.9%, 2.9%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [1.4%, 4.0%]   | 4     |
| Improvements ✅ <br /> (primary)   | -2.7% | [-2.7%, -2.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-2.7%, 2.9%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9f54abeebae58244d0d1c692af6f37d2b521e136&end=d1b36b73042e93fe4f006b06efda8d0018ccc5bd&stat=cycles:u)
<details>
<summary>Results (primary 2.1%, secondary 1.7%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1%  | [2.0%, 2.2%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.1%, 2.5%]   | 7     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | 2.1%  | [2.0%, 2.2%]   | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9f54abeebae58244d0d1c692af6f37d2b521e136&end=d1b36b73042e93fe4f006b06efda8d0018ccc5bd&stat=size:linked_artifact)
<details>
<summary>Results (primary 31.4%, secondary 33.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 31.4% | [0.0%, 112.6%] | 59    |
| Regressions ❌ <br /> (secondary)  | 33.6% | [0.0%, 120.4%] | 51    |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 31.4% | [0.0%, 112.6%] | 59    |
</details>

**Bootstrap**: 483.918s -> 491.944s (1.66%)
**Artifact size**: 390.81 MiB -> 390.86 MiB (0.01%)

bjorn3 commented:
Would adding a `requires_lto` check at https://github.com/rust-lang/rust/blob/2caa8fe4163e04b2df1c80f8024c346597fdc9e7/compiler/rustc_codegen_ssa/src/back/write.rs#L142 instead work? Your current code would cause object files containing both machine code and bitcode to be generated rather than standalone bitcode files, even when only the latter is necessary.

bjorn3 commented:
Yeah, I think so. Probably should update the docs for it though. It still says that it is mostly useful for Emscripten when Emscripten hasn't needed that ever since we got wasm.

Flakebi committed:
Fix requires_lto targets needing lto set in cargo

Targets that set `requires_lto = true` were not actually using lto when
compiling with cargo by default. They needed an extra `lto = true` in
`Cargo.toml` to work.

Fix this by letting lto take precedence over the `embed_bitcode` flag
when lto is required by a target.

If both these flags would be supplied by the user, an error is
generated. However, this did not happen when lto was requested by the
target instead of the user.

Flakebi commented:
I was too early claiming that it works, I tested some more (compiling amdgpu-rs examples) and it broke some of them (somehow it ignored CARGO_BUILD_RUSTFLAGS for obj_is_bitcode and EmitObj::Bitcode completely or in parts).
With `EmitObj::ObjectCode(BitcodeSection::Full)` everything seems to work.
So, this is pretty much the change from before now, but checking `requires_lto` instead of `lto != No`.

Flakebi commented:
@bjorn3, is this fix something you could get merged?

Would it be worth to have another rust-timer run before? (I see this is currently labelled with perf-regression but that comes from an earlier version of this PR.)

bjorn3 commented:
I figured @Mark-Simulacrum would approve given that they are assigned. I strongly doubt the current state will cause a perf regression.

@bors r+

Mark-Simulacrum commented:
For the future, I don't look at waiting-on-author PRs at all today, so I wouldn't have seen this. I'd recommend updating labels (e.g., `@rustbot ready`).

Flakebi commented:
Ups, I didn’t see the label was pointing to me, sorry. Thanks for approving!
